### PR TITLE
Admin Wallet Lock

### LINKED ISSUE #52: Adminwallet transactions should be synchronized
Use mongodb for atomic lock operations
Env variable for mongo url

Definition of Done
--
- [x] Nonce is incremented only for transactions sent successfully to the blockchain (ie got a tx hash)
- [x] Once tx hash is received lock is released and next server can try to grab lock
- [x] No race conditions exist, one server has lock at a time and nonce incremented sequentially
- [x] Unit tests to check the order

### PR DESCRIPTION
**Pull request by StanislavShevchenko**
closes #52

### CODE REVIEW

File: src/server/utils/tx-manager.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+          try {
+            tx = await nextTr.cb(walletNonce.nonce);
+          } catch (e) {
+            console.log(e);
+            await this.unlock(nextTr.address, walletNonce.nonce);
+          }
+          
+          await this.unlock(nextTr.address, walletNonce.nonce + 1);
+          
+          nextTr.done(tx)
```
Comment (sirpy) [inline src/server/utils/tx-manager.js]: caller won't know if tx failed or not as the "done" callback always uses "resolve" and never reject

Comment (Nordwhale) [reply src/server/utils/tx-manager.js]: ok. we will add reject callback

**Comment by Nordwhale** [comment]
> There is one main issue
> The semantics of sendTransaction and sendNative have changed, they no longer resolve and reject as previously done, so it will mess up anyone using them.
> 
> I would have preferred to see the same semantic of the mutex i've used before just mutex being a mongodb based mutex.
> 
> ```
> const {nonce} = await mongomutex.lock()
> try{
> <perform tx>
> unlock()
> }
> catch(e) {
> unlock(nonce)
> }
> ```
> 
> we could also use filters in watch to only get changes relevant to our address record and when the lock is set to false. or a tleast before calling this.run from watch check if lock===false and address===ouraddress
> example:
> 
> ```
> var filter = [{
>         $match: {
>             $and: [
>                 { "updateDescription.updatedFields.SomeFieldA": { $exists: true } },
>                 { operationType: "update" }]
>         }
>     }];
> 
>     var options = { fullDocument: 'updateLookup' };
>     db.collection('somecollection').watch(filter, options
> ```
> 
> i'll accept this as a valid solution to the test

about mutex. we don’t need them at all. because we use getAndUpdate. this operation lock the collection. so if run a lot of query they will add to queue in mongo. yesterday we've test it. run 20+ transaction in one time from different nodes. they are all successful completed one by one. but if it is important for u to use mutex we can change this part.

about watch filter. ok we will make changes

**Comment by sirpy** [comment]
> > There is one main issue
> > The semantics of sendTransaction and sendNative have changed, they no longer resolve and reject as previously done, so it will mess up anyone using them.
> > I would have preferred to see the same semantic of the mutex i've used before just mutex being a mongodb based mutex.
> > ```
> > const {nonce} = await mongomutex.lock()
> > try{
> > <perform tx>
> > unlock()
> > }
> > catch(e) {
> > unlock(nonce)
> > }
> > ```
> > 
> > 
> > we could also use filters in watch to only get changes relevant to our address record and when the lock is set to false. or a tleast before calling this.run from watch check if lock===false and address===ouraddress
> > example:
> > ```
> > var filter = [{
> >         $match: {
> >             $and: [
> >                 { "updateDescription.updatedFields.SomeFieldA": { $exists: true } },
> >                 { operationType: "update" }]
> >         }
> >     }];
> > 
> >     var options = { fullDocument: 'updateLookup' };
> >     db.collection('somecollection').watch(filter, options
> > ```
> > 
> > 
> > i'll accept this as a valid solution to the test
> 
> about mutex. we don’t need them at all. because we use getAndUpdate. this operation lock the collection. so if run a lot of query they will add to queue in mongo. yesterday we've test it. run 20+ transaction in one time from different nodes. they are all successful completed one by one. but if it is important for u to use mutex we can change this part.
> 
> about watch filter. ok we will make changes

I didnt mean mutex for mongo, i ment the same semantics i've used before with the await-mutex.
I ment that your implementation would be with same interface as a mutex.
i'm curious about the 20+ transactions you ran, what's the TPS rate? how quick was mongo to handle it?

**Comment by sirpy** [comment]
#52

**Comment by sirpy** [comment]
@StanislavShevchenko 
1. Add unit tests to test the definition of done, as this is a complicated and important feature.
2. regarding mongo, can we not use majorityReadConcern?

**Comment by StanislavShevchenko** [comment]
the type a semaphore is defined in the manager if the network number is not equal to LOCAL_NETWORK_ID, then type mongo semaphore. If you start the test and in an environment the fusi network is set, then will work through mongo and if there is a local network then works through mutex

**Pull request closed without merging**